main/main.go

```diff
@@ -227,136 +249,161 @@ func main() {
-				sourceBlockchain.RPCEndpoint.HTTPHeaders,
-				sourceBlockchain.RPCEndpoint.QueryParams,
-			)
-			if err != nil {
-				logger.Error(
-					"Failed to connect to node via RPC",
-					zap.String("blockchainID", sourceBlockchain.BlockchainID),
-					zap.Error(err),
-				)
-				return err
-			}
-
-			// Create the ApplicationRelayers
-			applicationRelayers, minHeight, err := createApplicationRelayers(
-				ctx,
-				logger,
-				metrics,
-				db,
-				ticker,
-				*sourceBlockchain,
-				network,
-				messageCreator,
-				&cfg,
-				ethClient,
-				destinationClients,
-			)
-			if err != nil {
-				logger.Error(
-					"Failed to create application relayers",
-					zap.String("blockchainID", sourceBlockchain.BlockchainID),
-					zap.Error(err),
-				)
-				return err
-			}
-			logger.Info(
-				"Created application relayers",
-				zap.String("blockchainID", sourceBlockchain.BlockchainID),
-			)
+			return relayer.ProcessManualWarpMessages(manualWarpMessages[sourceBlockchain.GetBlockchainID()])
+		})
 
+		// errgroup will cancel the context when the first goroutine returns an error
+		errGroup.Go(func() error {
 			// runListener runs until it errors or the context is cancelled by another goroutine
-			return runListener(
+			return relayer.RunListener(
 				ctx,
 				logger,
 				*sourceBlockchain,
-				health,
-				manualWarpMessages[blockchainID],
-				&cfg,
-				ethClient,
-				applicationRelayers,
-				minHeight,
+				sourceClients[sourceBlockchain.GetBlockchainID()],
+				isHealthy,
+				cfg.ProcessMissedBlocks,
+				minHeights[sourceBlockchain.GetBlockchainID()],
 			)
 		})
 	}
 	err = errGroup.Wait()
-	logger.Error(
-		"Relayer exiting.",
-		zap.Error(err),
-	)
+	logger.Error("Relayer exiting.", zap.Error(err))
 }
 
-// runListener creates a Listener instance and the ApplicationRelayers for a subnet.
-// The Listener listens for warp messages on that subnet, and the ApplicationRelayers handle delivery to the destination
-func runListener(
-	ctx context.Context,
+func createMessageHandlerFactories(
 	logger logging.Logger,
-	sourceBlockchain config.SourceBlockchain,
-	relayerHealth *atomic.Bool,
-	manualWarpMessages []*relayerTypes.WarpMessageInfo,
 	globalConfig *config.Config,
-	ethClient ethclient.Client,
-	applicationRelayers map[common.Hash]*relayer.ApplicationRelayer,
-	minHeight uint64,
-) error {
-	// Create the Listener
-	listener, err := relayer.NewListener(
-		logger,
-		sourceBlockchain,
-		relayerHealth,
-		globalConfig,
-		applicationRelayers,
-		minHeight,
-		ethClient,
-	)
-	if err != nil {
-		return fmt.Errorf("failed to create listener instance: %w", err)
+) (map[ids.ID]map[common.Address]messages.MessageHandlerFactory, error) {
+	messageHandlerFactories := make(map[ids.ID]map[common.Address]messages.MessageHandlerFactory)
+	for _, sourceBlockchain := range globalConfig.SourceBlockchains {
+		messageHandlerFactoriesForSource := make(map[common.Address]messages.MessageHandlerFactory)
```

Comment:
nit: can we rename to `messageProtocolHandlerFactory` or simiilar

Reply:
Rename which?

Reply:
`messageHandlerFactoriesForSource`